### DISCUSSION TITLE
Spanish Translation?

### DISCUSSION THREAD

**Discussion by TheEggo55** [chrislo27/PolyrhythmMania#3] [Ideas]
**What is this?**
It's simple, it's just a translation of the _Polyrhythm Mania_ fangame. This would allow more people to enjoy _Polyrhythm Mania_ easily, considering that a good portion of the RH community speaks Spanish natively, there is even a Spanish RHCord. Assuming that most edits would be made of text, the most conflicts would be that one of the fonts wouldn't have the characters with accents or tildes. Although I do not have an image at the moment, I do have an active development WIP fork called _Polirritmo Manía_, which should be based on some the the recent commits as of 8/22/2021.
_Edit 8/26/2021:_ The fork should be updated with the dev version 1.01 commit, and here is an image from this commit.
https://media.discordapp.net/attachments/834257030809387018/880642316833022002/unknown.png?width=767&height=431

**Comment by chrislo27** [comment]
The game is set up to be localized fairly easily but there's some issues with textures and the main Roboto font missing glyphs like you said.

I saw your localization fork of the game a few days ago and I think it's very neat. I'll reach out to you on Discord in the coming days regarding some other localization quirks such as internal consistency, making sure that when a default string gets updated that the other languages get updated as well, etc.

**Comment by chrislo27** [comment]
We should start by discussing special considerations that have to be made before starting a full localization process.

Here are some things I can think of off the top of my head, please reply and add some considerations you can think of as well.
* Language setting
  * There has to be a new setting for changing the language.
* Localized textures and other assets
  * The results "logos" for Try Again, OK, and Superb are textures and thus need a localized texture. Inherently the game also needs localized asset support.
  * As part of this, certain things/assets shouldn't be localized. For example, the name of the game shouldn't change.
* Tracking when strings need to be relocalized
  * It's easy to tell when a string isn't localized (use the `--log-missing-localizations` launch parameter), but hard to tell if, say, an English string gets updated and any translated strings will also have to be updated.
  * This is primarily an organization and bureaucratic issue.
* Fonts
  * The Roboto font doesn't have a whole lot of characters but it should have the right glyphs for Spanish. I recently updated the font for a more complete version for v1.1 ([example from Stack Exchange](https://spanish.stackexchange.com/a/2029)): 
![image](https://user-images.githubusercontent.com/6299069/131045199-10f525e7-cc43-4d16-8987-d2c766767ced.png)
  * The internal font system doesn't support falling back to a reliable font (like Noto Sans for example).
  * The Rodin font has a lot more characters and is the one used in RHRE primarily. It might be a good fallback but it is a wider font.
* Language variants
  * Language locales have a language, country, and variant (see [`langs.json`](https://github.com/chrislo27/PolyrhythmMania/blob/master/core/assets/localization/langs.json) or [RHRE's `localization` folder](https://github.com/chrislo27/RhythmHeavenRemixEditor/tree/dev/core/assets/localization))
  * Spanish is one of the most spoken languages worldwide, so there are a lot of country-specific locales and even variants within countries.
    * We have to make sure that this particular localization of Spanish has an appropriate region. For example, in a game like *Super Mario Odyssey*, there's a Spain ("Español (España)") version and Latin America ("Español (Latinoamérica)") version of Spanish.
    * I'm aware that the English version of Polyrhythm Mania is Canada-specific, so it may be a good exercise to also handle the country/variant versions for English.
  * It will be useful to have other native Spanish speakers peer review the localization and ensure it retains the same semantics as the English version.

**Reply by TheEggo55** [reply]
- Localized textures
   * [I have made textures for the Superb/OK/Try Again textures](https://github.com/TheEggo55/PolirritmoMania/tree/master/core/assets/textures/results/ranking), and looking into the files, it doesn't seem as though there will be much more textures to edit.
   * If there are more textures to edit, however, I will be open to edit them into the necessary language.
   * It is fine to not have the game name changed between languages.
- Language Variants
   * The Spanish I am using is mostly from what I've inherited from my family, and from what I've learned myself, which would be somewhat of a mix between Mexico and United States Spanish. Though to be on the safe side, I would prefer this language to be a United States ("Español (Estados Unidos)" or "Español (EE. UU.)") version.
   * Fortunately, there several Spanish-speaking people in the RH community that I would be able to talk to and review what I have, and a few that already help me check some stuff, like alternate language YouTube video titles/descriptions.

The rest of your points are completely valid, and I don't have anything else to say about them.

**Reply by chrislo27** [reply]
One thing I thought of was about how releases need to be handled with respect to localization. I don't want releases to be blocked on localization (RHRE used to have that issue), meaning I won't always delay a release waiting for languages to have their strings translated. This means that localizations may always be one step behind but that's a consequence of the approach I want to take.

As for checking what base strings have been changed, it could be as simple as doing a diff on the last update and then updating strings as necessary for a patch update. That would put the onus on the localizer for each language which I think is acceptable.

The language variant you want to use should be named [`es_US`](https://lh.2xlibre.net/locale/es_US/) then, for Spanish (United States). You can add another entry in `langs.json` to fit that accordingly. The .properties file should be named `default_es_US.properties`.

Localized asset support will have to be on a case-by-case basis. Results text sprites like you mentioned is definitely part of that.

I don't think I can really start working on localization integration until v1.2 since I have a lot of things to work on first for v1.1. However, I will try to have a new Language setting for v1.0.1 that will read `langs.json` and add the languages in a list accordingly.

If you're not already aware, you can easily test your strings on the *release builds* by adding directories and files alongside the launch script file for whichever assets you want to override for the platform agnostic version. Something like this screenshot:
![image](https://user-images.githubusercontent.com/6299069/132963075-e874fa2c-66bb-4487-864a-b4b03dcf0a83.png)

**Reply by User670** [reply]
> If you're not already aware, you can easily test your strings on the _release builds_ by adding directories and files alongside the PolyrhythmMania.jar file for whichever assets you want to override for the platform agnostic version. Something like this screenshot:

Uh, how, exactly? I placed a `localization` folder in the location the screenshot shows, and inside of it I placed a modified `langs.json` and a `zh-CN.properties`, game didn't recognize. I changed the .properties file to `default.properties` hoping it would replace the default English, also fail. Did I do something wrong, or did you remove this feature?

edit: I'm doing it on release 1.0.2

**Reply by chrislo27** [reply]
I neglected to add what the name of the .properties file should be in my previous reply so I have edited that. In your case, it should be `default_zh_CN.properties`.

I also updated my previous reply due to an error: the assets you want to override should be alongside where you're launching the game from. This means that if you're launching from the launch scripts, that these directories should be siblings of the script. If you are directly using a terminal and `cd`'d into the `bin` directory and launched it from the command line, then my previous example would've worked.

Please note that the game was only intended to be localized in a language that uses Latin script; there are no plans to change this. You may find issues with other scripts.

**Comment by User670** [comment]
Here are some things I can think of (or have encountered) regarding issues with localization

- ~~Han/CJK character variations. China Mainland, Taiwan and Japan use a shared set of Han characters, but each region write them slightly differently (although they use the same Unicode codepoints). For Chinese (China), Chinese (Taiwan) and Japanese, different fonts have to be used to allow displaying the different-shaped glyphs. (Annnnd there is the issue of "which one to use when my language is not Chinese or Japanese" lol)~~
- ~~Line breaking. I hope your text rendering engine automatically handles line breaking (word warping) for long texts because in most cases naive approaches don't work well.~~

Edit: Apparently the game is not intended to be localized into anything that doesn't use the Latin script. In this case, CJK isn;t an issue, and the naive approach to line breaking for Latin script works well for Latin script.

**Comment by GierkowiecTV** [comment]
on topic of a translation [https://www.mediafire.com/folder/skgll40y3w2s8/localization](url)
made my own polish translation 
![image](https://user-images.githubusercontent.com/87199096/134417490-e05a63e1-5517-4706-9be1-fa646827e997.png)
![image](https://user-images.githubusercontent.com/87199096/134417506-9f2c8312-45ef-424a-a242-6fa437a0b1d3.png)
![image](https://user-images.githubusercontent.com/87199096/134417540-41263652-04de-4185-8f29-765ceced76f1.png)
